### PR TITLE
Add methods to return the inner value for direction types

### PR DESCRIPTION
**Pull request by Multirious**
# Objective

Currently in order to retrieve the inner values from direction types is that you need to use the `Deref` trait or `From`/`Into`. `Deref` that is currently implemented is an anti-pattern that I believe should be less relied upon.
This pull-request add getters for retrieving the inner values for direction types.

Advantages of getters:
- Let rust-analyzer to list out available methods for users to understand better to on how to get the inner value. (This happens to me. I really don't know how to get the value until I look through the source code.)
- They are simple.
- Generally won't be ambiguous in most context. Traits such as `From`/`Into` will require fully qualified syntax from time to time.
- Unsurprising result.

Disadvantages of getters:
- More verbose

Advantages of deref polymorphism:
- You reduce boilerplate for getting the value and call inner methods by:
  ```rust
  let dir = Dir3::new(Vec3::UP).unwrap();
  // getting value
  let value = *dir;
  // instead of using getters
  let value = dir.vec3();

  // calling methods for the inner vector
  dir.xy();
  // instead of using getters
  dir.vec3().xy();
  ```

Disadvantages of deref polymorphism:
- When under more level of indirection, it will requires more dereferencing which will get ugly in some part:
  ```rust
  // getting value
  let value = **dir;
  // instead of using getters
  let value = dir.vec3();

  // calling methods for the inner vector
  dir.xy();
  // instead of using getters
  dir.vec3().xy();
  ```

[More detail here](https://rust-unofficial.github.io/patterns/anti_patterns/deref.html).

Edit: Update information for From/Into trait.
Edit: Advantages and disadvantages.

## Solution

Add `vec2` method for Dir2.
Add `vec3` method for Dir3.
Add `vec3a` method for Dir3A.

### CODE REVIEW

**Comment by Multirious** [comment]
I just realized that method is probably inaccessible unless specifying the type name because of the auto deref.

**Comment by afonsolage** [comment]
I'm not sure this is the best approach. There is already `From` implemented to cast from/into (I just noticed there is no `From<Dir2> for Vec2`), also it's possible to `Deref`, like you said.

Maybe the documentation can be improved to describe how to get the inner value?

**Comment by NthTensor** [comment]
It does seem like `From + Deref` is more idiomatic. `From<Dir2>` should **definitely** be implemented on `Vec2`. I'd suppose I wouldn't hate it if we had both `From` and access methods, but I'd like to hear more from the author as to why the proposed solution is also necessary.

Can you elaborate on your `rust-analyzer` comment?

**Comment by Multirious** [comment]
> It does seem like `From + Deref` is more idiomatic. `From<Dir2>` should **definitely** be implemented on `Vec2`. I suppose I wouldn't hate it if we had both `From` and access methods, but I'd like to hear more from the author as to why the proposed solution is also necessary.
> 
> Can you elaborate on your `rust-analyzer` comment?

Without the getter:
image

With the getter:
image
It's clear right away on how to access the `Vec2`.

**Comment by Multirious** [comment]
> Maybe the documentation can be improved to describe how to get the inner value?

Absolutely! As it stands, deref polymorphism is useful, but as detailed in [here](https://rust-unofficial.github.io/patterns/anti_patterns/deref.html), this is completely hidden from user. It should be clear for everyone that we're using the deref to access the value.

**Comment by afonsolage** [comment]
>  deref polymorphism is known as an anti-pattern

There is no polymorphism here, it's a new-type pattern. `DirX` isn't a subtype of `VecX` it is an extension/specialization. The `Deref` here is so that you can use any method in `DirX` as you would in `VecX`. The official docs of `Deref` [even mention ](https://doc.rust-lang.org/std/ops/trait.Deref.html#when-to-implement-deref-or-derefmut) this kind of usage:

> 1. a value of the type transparently behaves like a value of the target type;
> 2. the implementation of the deref function is cheap; and
> 3. users of the type will not be surprised by any deref coercion behaviour.

In our case, `DirX` is just a `VecX` with is guaranteed to be of unit length.

```rust
let d = Dir3::X;
let xz = d.xz(); // d acting like a `Vec3`
let vec3 = *d; // If a Vec3 is needed, just deref copy it
```

I've opened a [follow-up issue](https://github.com/bevyengine/bevy/issues/12529) to add the missing `From<Dir2> for Vec2`

**Comment by Multirious** [comment]
> There is no polymorphism here, it's a new-type pattern.

It's a new-type pattern that implements **deref polymorphism**.

> ```
> let xz = d.xz(); // d acting like a `Vec3`
> ```

This is exactly what deref polymorphism is.

Let's looks in the [deref polymorphism](https://rust-unofficial.github.io/patterns/anti_patterns/deref.html) example:
```rust
use std::ops::Deref;

struct Foo {}

impl Foo {
    fn m(&self) {
        //..
    }
}

struct Bar {
    f: Foo,
}

impl Deref for Bar {
    type Target = Foo;
    fn deref(&self) -> &Foo {
        &self.f
    }
}

fn main() {
    let b = Bar { f: Foo {} };
    b.m();
}
```

Here's the simplified implementation of Vec3 and Dir3:
```rust
use std::ops::Deref;

struct Vec3 {
    x: f32,
    y: f32,
    z: f32,
}

impl Vec3 {
    fn xz(&self) {
    }
}

struct Dir3(Vec3);

impl Deref for Dir3 {
    type Target = Vec3;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

fn main() {
    let d = Dir3(Vec3 { x: 1., y: 0., z: 0. });
    let xz = d.xz();
}
```
If you didn't notice, this looks exactly like the example from deref polymorphism.

**Comment by NthTensor** [comment]
Interesting discussion here. Removing our use of Deref would be controversial, so let's call that out of scope for this PR. I have been swayed a bit, and would be interested in seeing follow up. This "anti-pattern" is certainly common, but that doesn't mean it's good use here.

Regarding the proposed changes, I think discovery is worth some redundancy. I support having access methods in addition to Deref+From. Personally I would just call all of these into_vec.

**Comment by alice-i-cecile** [comment]
I'm lukewarm in favor of this if it's called `into_vec` and the methods are `const`.

**Comment by Jondolf** [comment]
I don't have strong opinions on the topic, but for some prior art:

Nalgebra's `Unit` types [implement `Deref`](https://docs.rs/nalgebra/latest/nalgebra/base/struct.Unit.html#impl-Deref-for-Unit%3CT%3E). They also have [`into_inner`](https://docs.rs/nalgebra/latest/nalgebra/base/struct.Unit.html#method.into_inner), which is equivalent to the proposed `vec2`/`vec3`/`vec3a`, but under a unified name.

Personally I consider our direction types to be Bevy's equivalents of Nalgebra's unit vectors and think of them primarily as vectors with extra restrictions. The direction types are newtypes of the vector types, and implementing `Deref` for newtypes is a very common pattern in Bevy (although perhaps for slightly different reasons in general), even though I guess in this case it could qualify under the mentioned deref polymorphism. Either way, imo the `Deref` impl makes at least some sense, but I'm fine with adding separate methods too.

For naming, Glam has methods like `vec2` to convert a `DVec2` to a `Vec2`, for example. So this proposal is in line with Glam in that sense. But a unified name like `into_vec` or `into_inner` could be fine too (except `into_vec` implies turning it into a `Vec<f32>`, so I wouldn't go for that)

**Comment by Multirious** [comment]
I like to follow the [naming guidelines](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv) generally.

I decided to use these names for the initial commit because:
- It's the most straightforward.
- Type is in the name. It's easily parsed without an LSP (e.g., reading a commit on the web).

I agree that using only `vec` could be confusing.

I will make the method const.
According to the naming guidelines, `into_` prefix should be use when the methods consumes self. Should I change to that?

**Comment by NthTensor** [comment]
Whoops, I missed that this takes borrowed self and does an implicit copy. to_vec2/3/3a seems like the way to go then, considering it would be best to avoid ambiguity with Vec<T>. Do I have that naming convention right?

**Comment by Jondolf** [comment]
I guess ownership-wise `to_` would be the correct prefix according to the guidelines, but they also say this:

> Conversions prefixed `to_`, on the other hand, typically stay at the same level of abstraction but do some work to change from one representation to another.

and `to_` is listed as "expensive". I mainly associate `to_` with methods that do some actual computations or transformations to change the underlying data to a different format, like how `to_radians` converts degrees to radians or how `to_lowercase` iterates through a `str` to construct a lowercase `String`. `to_vec2/3/3a` would be basically free and does not modify the underlying data in any way, and instead just gets the inner value stored in the direction type. So I'd personally prefer `as_vec2/3/3a`, if not for following guidelines then at least for consistency with Glam.

File: crates/bevy_math/src/direction.rs [hunk 1]
```diff
@@ -136,6 +136,11 @@ impl Dir2 {
     pub fn from_xy(x: f32, y: f32) -> Result<Self, InvalidDirectionError> {
         Self::new(Vec2::new(x, y))
     }
+
+    /// Returning the inner [`Vec2`]
```
Comment (Jondolf) [inline crates/bevy_math/src/direction.rs:140]: This comment format is quite strange. The verb should generally be in the present tense, not present continuous.

```suggestion
    /// Returns the inner [`Vec2`]
```

Comment (Multirious) [reply crates/bevy_math/src/direction.rs:140]: My bad. English is not my first language.

**Comment by Multirious** [comment]
https://stackoverflow.com/a/73861738
> A quicker way to think about it:
- if it consumes the data, use into_*
- if it returns another "view" of the data, use as_* - otherwise use to_*

**Comment by Jondolf** [comment]
I'll suggest once more that if we want a `something_vec2/3/3a` prefix, then I think we really should go for `as_*` for the sake of consistency with Glam regardless of what may or may not be technically more correct according to the guidance. It's confusing to have `as_dvec2` but to also have `to_vec2`. With `Deref` implemented, these are both exposed, and even if they weren't, I'd expect the naming to be consistent between `Vec2` and `Dir2`.

kuva

kuva

**Review by alice-i-cecile — APPROVED** [review]

**Review by NthTensor — APPROVED** [review]
Sorry this ended up with so much bike shedding.

**Review by james7132 — APPROVED** [review]

**Comment by Multirious** [comment]
My bad if I may have been a bit unclear on the explanation of things. This is my first actual PR attempt after all 😃. Just finnally gets to contribute to something I like.

The summary for adding a getters is that it would've been more ergonomic long-term because of its simple usage.

**Pull request merged**